sycl/include/sycl/marray.hpp

```diff
@@ -39,43 +39,30 @@ template <typename T, typename... Ts> struct GetMArrayArgsSize<T, Ts...> {
   static constexpr std::size_t value = 1 + GetMArrayArgsSize<Ts...>::value;
 };
 
-} // namespace detail
+// Trait for checking if an argument type is either convertible to the data
+// type or an array of types convertible to the data type.
+template <typename DataT, typename T>
+struct IsSuitableArgType : std::is_convertible<T, DataT> {};
+template <typename DataT, typename T, size_t N>
+struct IsSuitableArgType<DataT, marray<T, N>> : std::is_convertible<T, DataT> {
+};
 
-/// Provides a cross-platform math array class template that works on
-/// SYCL devices as well as in host C++ code.
-///
-/// \ingroup sycl_api
-template <typename Type, std::size_t NumElements> class marray {
-  using DataT = Type;
-
-public:
-  using value_type = Type;
-  using reference = Type &;
-  using const_reference = const Type &;
-  using iterator = Type *;
-  using const_iterator = const Type *;
+// Trait for computing the conjunction of of IsSuitableArgType. The empty type
+// list will trivially evaluate to true.
+template <typename DataT, typename... ArgTN>
+struct AllSuitableArgTypes
+    : std::conjunction<IsSuitableArgType<DataT, ArgTN>...> {};
 
+class HelperFlattenMArrayArg {
```

Comment:
Nit: I would probably have named it `FlattenMArrayArgHelper`, but it's a very small preference.